src/main/java/com/powsybl/caseserver/service/S3CaseService.java

```diff
@@ -0,0 +1,558 @@
+                f);
+    }
+
+    @Override
+    public String getFormat(UUID caseUuid) {
+        String format = getCaseMetadata(caseUuid).get(FORMAT_HEADER_KEY);
+        if (format == null) {
+            return withS3DownloadedTempPath(caseUuid, this::getFormat);
+        }
+        return format;
+    }
+
+    public String getCompressionFormat(UUID caseUuid) {
+        return caseMetadataRepository.findById(caseUuid).orElseThrow().getCompressionFormat();
+    }
+
+    public String getOriginalFilename(UUID caseUuid) {
+        return caseMetadataRepository.findById(caseUuid).orElseThrow().getOriginalFilename();
+    }
+
+    public Map<String, String> getCaseMetadata(UUID caseUuid) {
+        String caseFileKey = getCaseFileObjectKey(caseUuid);
+        HeadObjectResponse headObjectResponse = s3Client.headObject(HeadObjectRequest.builder()
+                .bucket(bucketName)
+                .key(caseFileKey)
+                .build());
+
+        return headObjectResponse.metadata();
+    }
+
+    // key format is "gsi-cases/UUID/filename"
+    private UUID parseUuidFromKey(String key) {
+        int firstSlash = key.indexOf(DELIMITER);
+        int secondSlash = key.indexOf(DELIMITER, firstSlash + 1);
+        return UUID.fromString(key.substring(firstSlash + 1, secondSlash));
+    }
+
+    private String parseFilenameFromKey(String key) {
+        int firstSlash = key.indexOf(DELIMITER);
+        int secondSlash = key.indexOf(DELIMITER, firstSlash + 1);
+        return key.substring(secondSlash + 1);
+    }
+
+    private String uuidToPrefixKey(UUID uuid) {
+        return CASES_PREFIX + uuid.toString() + DELIMITER;
+    }
+
+    private String uuidAndFilenameToKey(UUID uuid, String filename) {
+        return uuidToPrefixKey(uuid) + filename;
+    }
+
+    private List<S3Object> getCasesSummaries(String prefix) {
+        List<S3Object> s3Objects = new ArrayList<>();
+        ListObjectsV2Iterable listObjectsV2Iterable = s3Client.listObjectsV2Paginator(getListObjectsV2Request(prefix));
+        listObjectsV2Iterable.iterator().forEachRemaining(listObjectsChunk ->
+            s3Objects.addAll(listObjectsChunk.contents())
+        );
+        return s3Objects;
+    }
+
+    private ListObjectsV2Request getListObjectsV2Request(String prefix) {
+        return ListObjectsV2Request.builder().bucket(bucketName).prefix(prefix).build();
+    }
+
+    private List<S3Object> getCaseFileSummaries(UUID caseUuid) {
+        List<S3Object> files = getCasesSummaries(uuidToPrefixKey(caseUuid));
+        if (files.size() > 1) {
+            LOGGER.warn("Multiple files for case {}", caseUuid);
+        }
+        return files;
+    }
+
+    public String getCaseFileObjectKey(UUID caseUuid) {
+        List<String> allKeys = getCaseFileSummaries(caseUuid).stream()
+                .map(S3Object::key)
+                .toList();
+        return allKeys.stream()
+                .filter(key -> key.contains(".zip"))
+                .findFirst()
```

Comment:
should not use the first zip, it will fail if zip has nested zips

Reply:
This code is not relevant anymore. We can use the originalFilename

Reply:
I fixed it